### DISCUSSION TITLE
asyncio limited subset on Trinket M0

### DISCUSSION THREAD

**Discussion by markhildreth** [micropython/micropython#12556] [SAMD] [answered]
I downloaded MicroPython ([from here](https://micropython.org/download/ADAFRUIT_TRINKET_M0/)) onto my Adafruit Trinket M0, and was hoping to use asyncio. Unfortunately, I could not find the `asyncio` module.

```
MicroPython v1.20.0 on 2023-04-26; Trinket M0 with SAMD21E18A
Type "help()" for more information.
>>> import asyncio
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: module not found
>>> import uasyncio
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: module not found
```

However, I did find a `_uasyncio` module, although it was extremely limited in capability:

```
MicroPython v1.20.0 on 2023-04-26; Trinket M0 with SAMD21E18A
Type "help()" for more information.
>>> import _uasyncio as asyncio
>>> dir(asyncio)
['__class__', '__name__', '__dict__', 'Task', 'TaskQueue']
```

Am I missing something obvious? Does the port for this device not have a fully functional asyncio implementation?

**Comment by markhildreth** [comment]
So it looks  like the asyncio module is indeed not being built. The [`manifest.py`](https://github.com/micropython/micropython/blob/abb44694d85bc8e139b9969f6d033d9fd946e641/ports/samd/boards/manifest.py) for samd21 boards seems to not include asyncio, whereas the esp32 board's manifest.py contains this line:

https://github.com/micropython/micropython/blob/abb44694d85bc8e139b9969f6d033d9fd946e641/ports/esp32/boards/manifest.py#L2

I tried to change this locally and build using a samd21 manifest that looked like this:

```
freeze("$(PORT_DIR)/modules")
include("$(MPY_DIR)/extmod/asyncio")
```

During build, however, I received this error message:

```
$ make -C ports/samd BOARD=ADAFRUIT_TRINKET_M0
(...snip...)
LINK build-ADAFRUIT_TRINKET_M0/firmware.elf
arm-none-eabi-ld: build-ADAFRUIT_TRINKET_M0/firmware.elf section `.text' will not fit in region `FLASH'
arm-none-eabi-ld: region `FLASH' overflowed by 5032 bytes
Memory region         Used Size  Region Size  %age Used
           FLASH:      193448 B       184 KB    102.67%
             RAM:        1880 B        32 KB      5.74%
make: *** [Makefile:201: build-ADAFRUIT_TRINKET_M0/firmware.elf] Error 1
```

I understand this as there just isn't enough flash memory to store the micropython installation that includes asyncio. Is my understanding correct?

**Reply by robert-hh** [reply, marked as answer]
Yes, you are right. You could increase the size for the code. Then, the file system will be smaller. The code size is a setting in mpcongigboard.mk, or you can define it at the command line during build. it is MICROPY_HW_CODESIZE.

**Reply by robert-hh** [reply]
At the moment it is set to 184k. so with 200k you get 26k more for code.

**Comment by jimmo** [comment]
> However, I did find a _uasyncio module, although it was extremely limited in capability:

@markhildreth Just to follow up on this point -- `_asyncio` is a built-in helper to optimise some parts of `asyncio`, and probably we shouldn't include it if not freezing the actual `asyncio` module.

But it does mean that you could actually install asyncio to the filesystem.
